demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <functional>
+#include <list>
+#include <map>
+#include <stdexcept>
+#include <string>
+
+#include "utils_gapi/backend_description.hpp"
+
+#ifdef GAPI_ONNX_BACKEND
+#include <opencv2/gapi/infer/onnx.hpp>
+
+const std::initializer_list<std::string> getONNXSupportedEP();
+
+template<class ExecNetwork, class ...Args>
+void applyONNXProviders(ExecNetwork& net, const ModelConfig &config, inference_backends_t backend_cfgs = inference_backends_t{});
+
+template<class ExecNetwork>
+struct BackendApplicator<ExecNetwork,
+                         cv::gapi::onnx::Params> {
+    static cv::gapi::GNetPackage apply(const std::string &model_path, const ModelConfig & config, const inference_backends_t &backends) {
+        auto net =
+            cv::gapi::onnx::Params<ExecNetwork>{
+                model_path
+            };
+        applyONNXProviders(net, config, backends);
+        return cv::gapi::networks(net);
+    }
+};
+
+namespace {
+template<class ExecNetwork>
+struct ProviderApplicator {
+    static void apply(ExecNetwork &, const ModelConfig &, const inference_backends_t &backends) {
+        if (backends.size() > 1) {
+            throw std::runtime_error("ONNX CPU execution provider must be only provider in the provider list or "
+                                     "be the last in order");
+        }
+
+        const BackendDescription &backend = backends.front();
+        if (backend.properties.size() >= 2) {
+            throw std::runtime_error("ONNX CPU execution provider doesn't support any arguments");
+        }
+    }
+};
+
+#ifdef GAPI_ONNX_BACKEND_EP_EXTENSION
+template<class ExecNetwork, class Provider>
+struct ProviderApplicator {
+    static void apply(ExecNetwork &net, const ModelConfig &config, const inference_backends_t &backends) {
+        std::string ep_selected_device = config.deviceName;
+        const BackendDescription &backend = backends.front();
+        if (backend.properties.size() >= 2) {
+            ep_selected_device = backend.properties[1]; // second arg for ONNX is device
+        }
+        net.cfgAddExecutionProvider(Provider(ep_selected_device));
+    }
+};
```

Comment:
I can't compile it with enabled `#define GAPI_ONNX_BACKEND_EP_EXTENSION`. The error is
`C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(63,2): error C2977: '`anonymous-namespace'
::ProviderApplicator': too many template arguments [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(54,8): message : see declaration of '`anon
ymous-namespace'::ProviderApplicator' [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(76,2): error C2953: '`anonymous-namespace'
::ProviderApplicator': class template has already been defined [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_ga
pi.vcxproj]
C:\Users\vzlobin\r\z\demos\common\cpp_gapi\utils_gapi\include\utils_gapi/onnx_backend.hpp(66): message : see declaration of '`anonym
ous-namespace'::ProviderApplicator' [C:\Users\vzlobin\r\z\demos\build\common\cpp_gapi\utils_gapi\utils_gapi.vcxproj]`

Reply:
sorry, my fault. 
I'll fix that

Reply:
checked it with different MACROs, it must be OK now